### PR TITLE
PrestaShopCollection component

### PR DESCRIPTION
**Pull request by thomasnares**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x
| Description?  | Document the PrestaShopCollection component

### CODE REVIEW

File: development/components/database/prestashopcollection.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
+---
+title: PrestaShopCollection class
+---
+
+# The PrestaShopCollection component
+
+## Introduction
+
+The PrestaShopCollection component is a Collection of ObjectModel objects. It implements common useful `php` interfaces:
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
The PrestaShopCollection component is a Collection of ObjectModel objects. It implements common useful `PHP` interfaces: 
```

File: development/components/database/prestashopcollection.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+use PrestaShopCollection;
+
+$productCollection = (new PrestaShopCollection('Product'))
+    ->where('on_sale', '=', true)
+    ->orderBy('reference', 'desc')
+    ->setPageSize(100);
+
+if(count($productCollection) > 0){
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
if (count($productCollection) > 0) {
```

PSR-2 ;d

File: development/components/database/prestashopcollection.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    foreach($productCollection as $product){
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
    foreach ($productCollection as $product) {
```

File: development/components/database/prestashopcollection.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        // do something with your product
+    }
+}
+```
+
+## Create a PrestaShopCollection
+
+To create a PrestaShopCollection, simply instanciate a PrestaShopCollection with the ObjectModel type as constructor parameter.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
To create a PrestaShopCollection, simply instantiate a PrestaShopCollection with the ObjectModel type as a constructor parameter. 
```

File: development/components/database/prestashopcollection.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+To create a PrestaShopCollection, simply instanciate a PrestaShopCollection with the ObjectModel type as constructor parameter. 
+
+```php
+use PrestaShopCollection;
+
+$productCollection = new PrestaShopCollection('Product');
+```
+
+A second parameter can be used, to set the id of the lang to use (`$id_lang`), set at null by default.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
A second parameter is available to set the ID of the language to use (`$id_lang`). It is `null` by default.
```

File: development/components/database/prestashopcollection.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+use PrestaShopCollection;
+
+$idLang=1;
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
$idLang = 1;
```

File: development/components/database/prestashopcollection.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+$productCollection = new PrestaShopCollection('Product', $idLang);
+```
+
+{{% notice note %}}
+Using `$id_lang` parameter will allow to set a Lang Context when querying / filtering multilang fields. If not set, multilang fields will be returned as `array` in the Collection.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
Using `$id_lang` parameter will allow setting a Context language when querying/filtering multi-lang fields. If not set, multi-lang fields will be returned as an `array` in the Collection.
```

File: development/components/database/prestashopcollection.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+{{% /notice %}}
+
+## Get results, count results
+
+When your collection is created, you may need to get its content. Several methods are available to retrieve its content:
+
+- `getFirst()`: gets first item of the Collection
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
- `getFirst()`: gets the first item of the Collection
```

File: development/components/database/prestashopcollection.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+- `getLast()`: gets last item of the Collection
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
- `getLast()`: gets the last item of the Collection
```

File: development/components/database/prestashopcollection.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+- or simply access items via index, or iterate with `foreach`
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
- you can also access items via their index key, or iterate with `foreach`
```

File: development/components/database/prestashopcollection.md [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+$productCollection = new PrestaShopCollection('Product');
+
+$firstProduct = $productCollection->getFirst();
+$lastProduct = $productCollection->getLast();
+
+// you can access items in the collection by index, like a regular array (the class implements ArrayAccess). Index starts at 0
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
// you can access items in the collection by index, like in a regular array (this class implements ArrayAccess). The index starts at 0
```

File: development/components/database/prestashopcollection.md [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+$thirdProduct = $productCollection[2]; 
+
+// The collection is also iterable (the class implements Iterator): 
+foreach($productCollection as $product){
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
foreach ($productCollection as $product) {
```

File: development/components/database/prestashopcollection.md [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+$productCollection = new PrestaShopCollection('Product');
+$productCollection->where('on_sale', '=', true); // find products on sale
+```
+
+```php
+$productCollection = new PrestaShopCollection('Product');
+$productCollection->where('reference', 'LIKE', 'REF-%'); // find products with reference beginning by "REF-"
+```
+
+### Having method
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
### `Having` method
```

File: development/components/database/prestashopcollection.md [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+
+Another method is available, `having()`, which calls `where()` with the parameter `$method` set to `having`.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
There is also `having()` method available, which calls `where()` with the parameter `$method` set to `having`. 
```

File: development/components/database/prestashopcollection.md [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+By default, a `LEFT JOIN` will be used. `INNER JOIN` or `LEFT OUTER JOIN` are also available using the third parameter:
+
+```php
+$productCollection->join('categories', 'id_category', PrestaShopCollection::LEFT_JOIN);
+$productCollection->join('categories', 'id_category', PrestaShopCollection::INNER_JOIN);
+$productCollection->join('categories', 'id_category', PrestaShopCollection::LEFT_OUTER_JOIN);
+```
+
+### Using where with join
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
### Using `where` with `join`
```

File: development/components/database/prestashopcollection.md [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+
+## Ordering a Collection (order by)
+
+To order your collection, use the `orderBy()` method. 
+
+```php
+public function orderBy($field, $order = 'asc')
+```
+
+Ordering can be done in ascending or descending direction, with `asc` or `desc` `$order` parameter.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
Ordering can be done in ascending or descending directions, with `asc` or `desc` in the `$order` parameter.
```

File: development/components/database/prestashopcollection.md [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+$productCollection = (new PrestaShopCollection('Product'))
+    ->orderBy('reference', 'desc');
+```
+
+## Grouping (group by)
+
+To group your collection items on a field, use the `groupBy()` method.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
If you want to organize your collection items based on a specific field, you can make use of the `groupBy()` method.
```

File: development/components/database/prestashopcollection.md [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+```
+
+```php
+$productCollection = (new PrestaShopCollection('Product'))
+    ->groupBy('id_supplier');
+```
+
+## Paginating a Collection (offset / limit)
+
+When fetching large quantities of items, you may need to paginate your collection, with SQL `OFFSET / LIMIT` equivalent mecanismes.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
If you need to retrieve a large number of items, it may be necessary to use pagination techniques such as SQL's `OFFSET/LIMIT`. This will help you manage your collection more efficiently.
```

File: development/components/database/prestashopcollection.md [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+```php
+$productCollection = (new PrestaShopCollection('Product'))
+    ->groupBy('id_supplier');
+```
+
+## Paginating a Collection (offset / limit)
+
+When fetching large quantities of items, you may need to paginate your collection, with SQL `OFFSET / LIMIT` equivalent mecanismes. 
+
+To do so, use those two methods:
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
To do so, use these two methods:
```

File: development/components/database/prestashopcollection.md [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+public function setPageNumber($page_number)
+public function setPageSize($page_size)
+```
+
+To use this feature, first, set the $page_size:
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
First, set the $page_size:
```

File: development/components/database/prestashopcollection.md [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+
+```php
+$productCollection = (new PrestaShopCollection('Product'))
+    ->setPageSize(100); // will get only 100 items
+```
+
+If only `$page_size` is set, the collection will return the first `$page_size` items.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
If `$page_size` is set, the collection will return the first `$page_size` items. 
```

File: development/components/database/prestashopcollection.md [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+To use this feature, first, set the $page_size:
+
+```php
+$productCollection = (new PrestaShopCollection('Product'))
+    ->setPageSize(100); // will get only 100 items
+```
+
+If only `$page_size` is set, the collection will return the first `$page_size` items. 
+
+To do a pagination, and get the page 2 of the collection, use the following example:
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]:
```suggestion
If you need to paginate and access the second page of a collection, you can use the following example:
```

File: development/components/database/prestashopcollection.md [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+    `LIMIT 100` or `LIMIT 100 OFFSET 0` or `LIMIT 0,100`
+
+- Getting the page: `2` of a collection with a page_size of: `100` is the `PrestaShopCollection` equivalent of this `SQL` statement: 
+    `LIMIT 100 OFFSET 100` or `LIMIT 100,100`
+
+- Getting the page: `3` of a collection with a page_size of: `100` is the `PrestaShopCollection` equivalent of this `SQL` statement: 
+    `LIMIT 100 OFFSET 200` or `LIMIT 200,100`
+
+{{% notice info %}}
+Known issue: When not using `$id_lang`, and using pagination (`setPageSize()`, `setPageNumber()`), pagination will be broken. Please use a Lang Context if the entity is multi-lang enabled.
```
Comment (kpodemski) [inline development/components/database/prestashopcollection.md]: I don't understand this part now. I think you were more clear when we discussed it. Pagination is broken when getting multi-lang entries with `$id_lang` provided. Yes? Because you wrote:
`Please use a Lang Context if the entity is multi-lang enabled. `

**Review by kpodemski — CHANGES_REQUESTED** [review]
Well done, Thomas! The review is ready.

File: development/components/database/prestashopcollection.md (hunk 23 above)
Comment (thomasnares) [reply development/components/database/prestashopcollection.md]: ok, let me rephrase it

File: development/components/database/prestashopcollection.md [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
```
Comment (thomasnares) [inline development/components/database/prestashopcollection.md]:
```suggestion
Known issue: If the entity has multi-lang attributes, when not using `$id_lang` in the `PrestaShopCollection` constructor, and using pagination (`setPageSize()`, `setPageNumber()`), pagination will be broken. 
Please set an `$id_lang` in the constructor to use pagination features for entities with multi-lang attributes.
```

Comment (thomasnares) [reply development/components/database/prestashopcollection.md]: @kpodemski is it better?

Comment (thomasnares) [reply development/components/database/prestashopcollection.md]: To reproduce the issue, with the default PrestaShop product demo collection: 

```php
$productCollection = (new PrestaShopCollection('Product'))->setPageSize(10)->setPageNumber(1); 
var_dump(count($productCollection));
if(count($productCollection) > 0){
  foreach($productCollection as $product){
    var_dump($product->id, $product->name); 
  }
}
```

```
int(5)
int(1)
array(2) {
  [1]=>
  string(27) "Hummingbird printed t-shirt"
  [2]=>
  string(27) "Hummingbird printed t-shirt"
}
int(2)
array(2) {
  [1]=>
  string(27) "Hummingbird printed sweater"
  [2]=>
  string(27) "Hummingbird printed sweater"
}
int(3)
array(2) {
  [1]=>
  string(38) "The best is yet to come' Framed poster"
  [2]=>
  string(38) "The best is yet to come' Framed poster"
}
int(4)
array(2) {
  [1]=>
  string(34) "The adventure begins Framed poster"
  [2]=>
  string(34) "The adventure begins Framed poster"
}
int(5)
array(2) {
  [1]=>
  string(33) "Today is a good day Framed poster"
  [2]=>
  string(33) "Today is a good day Framed poster"
}
```

```php
$productCollection = (new PrestaShopCollection('Product', 1))->setPageSize(10)->setPageNumber(1); 
var_dump(count($productCollection));
if(count($productCollection) > 0){
  foreach($productCollection as $product){
    var_dump($product->id, $product->name); 
  }
}
```

```
int(10)
int(1)
string(27) "Hummingbird printed t-shirt"
int(2)
string(27) "Hummingbird printed sweater"
int(3)
string(38) "The best is yet to come' Framed poster"
int(4)
string(34) "The adventure begins Framed poster"
int(5)
string(33) "Today is a good day Framed poster"
int(6)
string(27) "Mug The best is yet to come"
int(7)
string(24) "Mug The adventure begins"
int(8)
string(23) "Mug Today is a good day"
int(9)
string(20) "Mountain fox cushion"
int(10)
string(18) "Brown bear cushion"
```

**Review by kpodemski — APPROVED** [review]
